src/btree/bt_delete.c

```diff
@@ -284,101 +297,92 @@ __tombstone_update_alloc(
     WT_PAGE *page;
-    WT_PAGE_DELETED *page_del;
     WT_ROW *rip;
     WT_TIME_WINDOW tw;
     WT_UPDATE **upd_array, **update_list, *upd;
     size_t size, total_size;
     uint32_t count, i;
 
-    btree = S2BT(session);
+    /*
+     * An operation is accessing a "deleted" page, and we're building an in-memory version of the
+     * page (making it look like all entries in the page were individually updated by a remove
+     * operation). We end up here if a transaction used a truncate call to delete the page without
+     * reading it, and something else that can't yet see the truncation decided to read the page.
+     *
+     * This can happen after the truncate transaction resolves, but it can also happen before. In
+     * the latter case, we need to keep track of the updates we populate the page with, so they can
+     * be found when the transaction resolves. The page we're loading might split, in which case
+     * finding the updates any other way would become a problem.
+     *
+     * The page_del structure passed in is either ref->ft_info.del, or under certain circumstances
+     * when that's unavailable, one extracted from the parent page's address cell.
+     */
+
     page = ref->page;
-    page_del = NULL;
     update_list = NULL;
 
+    /* For now fast-truncate is only supported for row-store. */
+    WT_ASSERT(session, page->type == WT_PAGE_ROW_LEAF);
+
     WT_STAT_CONN_DATA_INCR(session, cache_read_deleted);
 
     /* Track the prepared, fast-truncate pages we've had to instantiate. */
-    if (ref->ft_info.del != NULL && ref->ft_info.del->prepare_state != WT_PREPARE_INIT)
+    if (page_del != NULL && page_del->prepare_state != WT_PREPARE_INIT)
         WT_STAT_CONN_DATA_INCR(session, cache_read_deleted_prepared);
 
     /*
-     * Give the page a modify structure. If the tree is already dirty and so will be written, mark
-     * the page dirty. (We want to free the deleted pages, but if the handle is read-only or if the
-     * application never modifies the tree, we're not able to do so.)
+     * Give the page a modify structure and mark the page dirty if the tree isn't read-only. If the
+     * tree can be written, the page must be marked dirty: otherwise it can be discarded, and that
+     * will lose the truncate information if the parent page hasn't been reconciled since the
+     * truncation happened.
+     *
+     * If the tree cannot be written (checked in page-modify-set), we won't dirty the page. In this
+     * case the truncate information must have been read from the parent page's on-disk cell, so we
+     * can fetch it again if we discard the page and then reread it.
+     *
+     * Truncates can appear in read-only trees (whether a read-only open of the live database or via
+     * a checkpoint cursor) if they were not yet globally visible when the tree was checkpointed.
      */
     WT_RET(__wt_page_modify_init(session, page));
-    if (btree->modified)
-        __wt_page_modify_set(session, page);
```

Comment:
The comment is confusing. We are not checking read only here?

Reply:
The first thing __wt_page_modify_set does is check for a readonly tree and return without doing anything more, so an explicit check is redundant. Hence the comment, but I guess it's not good enough -- any suggestions for improving it?

Reply:
ok, got it. I would say probably rename the function `__wt_page_modify_set ` but I guess it's not a big deal. All good.